05-AdvancedTopics/web-search-mcp/README.md

```diff
@@ -0,0 +1,214 @@
+# Web Search MCP Server
+
+This advanced sample demonstrates how to build a sophisticated MCP server and client using FastMCP to extend LLM capabilities with real-time web data via SerpAPI, a critical skill for developing dynamic AI agents.
+
+## Overview
+
+This implementation features four tools that showcase MCP's ability to handle diverse, external API-driven tasks securely and efficiently:
+
+- **general_search**: For broad web results
+- **news_search**: For recent headlines
+- **product_search**: For e-commerce data
+- **qna**: For question-and-answer snippets
+
+## Features
+
+- **External API Integration**: Demonstrates secure handling of API keys and external requests
+- **Structured Data Parsing**: Shows how to transform API responses into LLM-friendly formats
+- **Error Handling**: Robust error handling with appropriate logging
+- **Interactive Client**: Includes both automated tests and an interactive mode for testing
+- **Context Management**: Leverages MCP Context for logging and tracking requests
+
+## Prerequisites
+
+- Python 3.8 or higher
+- SerpAPI API Key (Sign up at [SerpAPI](https://serpapi.com/) - free tier available)
+
+## Installation
+
+1. Install dependencies using uv (recommended) or pip:
+
+```bash
+# Using uv (recommended)
+uv pip install -r requirements.txt
+
+# Using pip
+pip install -r requirements.txt
+```
+
+2. Create a `.env` file in the project root with your SerpAPI key:
+
+```
+SERPAPI_KEY=your_serpapi_key_here
+```
+
+## Usage
+
+### Running the Server
+
+Start the MCP server:
+
+```bash
+python server.py
+```
+
+The server will run as a stdio-based MCP server that the client can connect to directly.
+
+### Running the Client
+
+Run the automated tests (this will automatically start the server):
+
+```bash
+python client.py
```

Comment:
Same comment as with the server, can you introduce the client first in a previous section and say what it does, this shouldn't be the first time the reader hears about client.py